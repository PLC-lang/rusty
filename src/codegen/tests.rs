// Copyright (c) 2020 Ghaith Hachem and Mathias Rieder
mod code_gen_tests;
mod codegen_error_messages_tests;
mod compare_instructions_tests;
mod constants_tests;
mod debug_tests;
mod directaccess_test;
mod expression_tests;
mod function_tests;
mod generics_test;
mod initialization_test;
mod parameters_tests;
mod statement_codegen_test;
mod string_tests;
#[cfg(feature = "verify")]
mod switch_case_tests;
mod typesystem_test;
