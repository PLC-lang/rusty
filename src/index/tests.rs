// Copyright (c) 2020 Ghaith Hachem and Mathias Rieder
mod generic_tests;
mod index_tests;
mod instance_resolver_tests;
