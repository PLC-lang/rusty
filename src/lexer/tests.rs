/// Copyright (c) 2020 Ghaith Hachem and Mathias Rieder

use pretty_assertions::{assert_eq, assert_ne};
#[test]
fn pou_tokens() {
    let mut lexer = super::lex("PROGRAM END_PROGRAM FUNCTION END_FUNCTION FUNCTION_BLOCK END_FUNCTION_BLOCK");
    assert_eq!(lexer.token, super::Token::KeywordProgram);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordEndProgram);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordFunction);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordEndFunction);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordFunctionBlock);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordEndFunctionBlock);
    lexer.advance();
}

#[test]
fn var_tokens() {
    let mut lexer = super::lex("VAR VAR_INPUT VAR_GLOBAL END_VAR");
    assert_eq!(lexer.token, super::Token::KeywordVar);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordVarInput);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordVarGlobal);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordEndVar);
}

#[test]
fn hello_is_an_identifier() {
    let mut lexer = super::lex("hello a12 _a12");
    assert_eq!(lexer.token, super::Token::Identifier, "{}", lexer.slice());
    lexer.advance();
    assert_eq!(lexer.token, super::Token::Identifier, "{}", lexer.slice());
    lexer.advance();
    assert_eq!(lexer.token, super::Token::Identifier, "{}", lexer.slice());
    lexer.advance();
}

#[test]
fn an_identifier_cannot_start_with_a_number() {
    let lexer = super::lex("2g12");
    assert_ne!(lexer.token, super::Token::Identifier);
}

#[test]
fn punctuations() {
    let lexer = super::lex(":");
    assert_eq!(lexer.token, super::Token::KeywordColon, "{}", lexer.slice());
    let lexer = super::lex(";");
    assert_eq!(
        lexer.token,
        super::Token::KeywordSemicolon,
        "{}",
        lexer.slice()
    );
}

#[test]
fn parens() {
    let mut lexer = super::lex("( )");
    assert_eq!(lexer.token, super::Token::KeywordParensOpen);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordParensClose);
}

#[test]
fn a_assignment_is_keword_assignment() {
    let lexer = super::lex(":=");
    assert_eq!(lexer.token, super::Token::KeywordAssignment);
}

#[test]
fn comma(){
    let lexer = super::lex(",");
    assert_eq!(lexer.token, super::Token::KeywordComma);
}

#[test]
fn operator_test() {
    let mut lexer = super::lex("+ - * / MOD = <> < > <= >=");
    assert_eq!(lexer.token, super::Token::OperatorPlus);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::OperatorMinus);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::OperatorMultiplication);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::OperatorDivision);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::OperatorModulo);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::OperatorEqual);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::OperatorNotEqual);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::OperatorLess);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::OperatorGreater);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::OperatorLessOrEqual);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::OperatorGreaterOrEqual);
}

#[test]
fn boolean_expression_test() {
    let mut lexer = super::lex("AND XOR OR NOT");
    assert_eq!(lexer.token, super::Token::OperatorAnd);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::OperatorXor);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::OperatorOr);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::OperatorNot);
}

#[test]
fn int_literals_test() {
    let mut lexer = super::lex("1 2 3 0123 321");

    for x in 0..5 {
        print!("{}", x);
        assert_eq!(lexer.token, super::Token::LiteralInteger);
        lexer.advance();
    }
}

#[test]
fn real_literals_test() {
    let mut lexer = super::lex("1.234 0.9E10");

        assert_eq!(lexer.token, super::Token::LiteralInteger);
        lexer.advance();
        assert_eq!(lexer.token, super::Token::KeywordDot);
        lexer.advance();
        assert_eq!(lexer.token, super::Token::LiteralInteger);
        lexer.advance();
        assert_eq!(lexer.token, super::Token::LiteralInteger);
        lexer.advance();
        assert_eq!(lexer.token, super::Token::KeywordDot);
        lexer.advance();
        assert_eq!(lexer.token, super::Token::LiteralInteger);
        lexer.advance();
        assert_eq!(lexer.token, super::Token::LiteralExponent);
        lexer.advance();
}

#[test]
fn a_full_program_generates_correct_token_sequence() {
    let mut lexer = super::lex(
        r"
        PROGRAM hello
        VAR
          a : INT;
        END_VAR
        END_PROGRAM
        ",
    );

    assert_eq!(lexer.token, super::Token::KeywordProgram);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::Identifier);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordVar);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::Identifier);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordColon);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::Identifier);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordSemicolon);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordEndVar);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordEndProgram);
}

#[test]
fn boolean_literals() {
    let mut lexer = super::lex(r" TRUE FALSE");
    assert_eq!(lexer.token, super::Token::LiteralTrue);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::LiteralFalse);
}

#[test]
fn if_expression() {
    let mut lexer = super::lex(
        r"
        IF THEN ELSIF ELSE END_IF
        ",
    );

    assert_eq!(lexer.token, super::Token::KeywordIf);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordThen);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordElseIf);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordElse);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordEndIf);
}

#[test]
fn for_statement() {
    let mut lexer = super::lex(
        r"
        FOR TO BY DO END_FOR
        ",
    );

    assert_eq!(lexer.token, super::Token::KeywordFor);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordTo);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordBy);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordDo);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordEndFor);
}

#[test]
fn while_statement() {
    let mut lexer = super::lex(
        r"
        WHILE DO END_WHILE
        ",
    );

    assert_eq!(lexer.token, super::Token::KeywordWhile);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordDo);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordEndWhile);
}

#[test]
fn repeat_statement() {
    let mut lexer = super::lex(
        r"
        REPEAT UNTIL END_REPEAT
        ",
    );

    assert_eq!(lexer.token, super::Token::KeywordRepeat);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordUntil);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordEndRepeat);
}

#[test]
fn case_statement() {
    let mut lexer = super::lex(
        r"
        CASE OF ELSE END_CASE
        ",
    );

    assert_eq!(lexer.token, super::Token::KeywordCase);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordOf);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordElse);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordEndCase);
}

#[test]
fn dot_statements() {
    let mut lexer = super::lex(
        r".. .",
    );

    assert_eq!(lexer.token, super::Token::KeywordDotDot);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordDot);
    lexer.advance();
}

#[test]
fn range_statements() {
    let mut lexer = super::lex(
        r"123..ABC",
    );

    println!("{:?}", lexer.token);
    lexer.advance();
    println!("{:?}", lexer.token);
    lexer.advance();
    println!("{:?}", lexer.token);
    lexer.advance();
}

#[test]
fn struct_enum_datatype() {
    let mut lexer = super::lex(
        r"TYPE STRUCT END_STRUCT END_TYPE",
    );

    assert_eq!(lexer.token, super::Token::KeywordType);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordStruct);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordEndStruct);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordEndType);
}

#[test]
fn array_parsing() {
    let mut lexer = super::lex(
        r"ARRAY OF x[5]",
    );

    assert_eq!(lexer.token, super::Token::KeywordArray);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordOf);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::Identifier);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordSquareParensOpen);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::LiteralInteger);
    lexer.advance();
    assert_eq!(lexer.token, super::Token::KeywordSquareParensClose);
    lexer.advance();
    
}

#[test]
fn string_parsing() {
    let mut lexer = super::lex(
        r"'AB C' 'AB$$' 'AB$''"
    );

    assert_eq!(lexer.token, super::Token::LiteralString);
    assert_eq!("'AB C'",lexer.slice());
    lexer.advance();
    assert_eq!(lexer.token, super::Token::LiteralString);
    assert_eq!("'AB$$'",lexer.slice());
    lexer.advance();
    assert_eq!(lexer.token, super::Token::LiteralString);
    assert_eq!("'AB$''",lexer.slice());
    lexer.advance();
}
    
