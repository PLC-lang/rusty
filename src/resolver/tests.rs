use crate::{
    ast::{self, CompilationUnit},
    index::{self, Index},
    lexer::lex,
};

use super::{AnnotationMap, TypeAnnotator};

#[cfg(test)]
mod const_resolver_tests;
#[cfg(test)]
mod resolve_control_statments;
#[cfg(test)]
mod resolve_expressions_tests;
#[cfg(test)]
mod resolve_literals_tests;

fn parse(src: &str) -> (CompilationUnit, Index) {
    let (mut unit, _) = crate::parser::parse(lex(src));

    ast::pre_process(&mut unit);
    let index = index::visitor::visit(&unit);
    (unit, index)
}

fn annotate(parse_result: &CompilationUnit, index: &Index) -> AnnotationMap {
    TypeAnnotator::visit_unit(index, parse_result)
}
