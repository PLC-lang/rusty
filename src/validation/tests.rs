// Copyright (c) 2020 Ghaith Hachem and Mathias Rieder
mod array_access_validation_test;
mod bitaccess_validation_test;
mod literals_validation_tests;
mod reference_resolve_tests;
mod variable_validation_tests;

mod statement_validation_tests;

mod generic_validation_tests;
