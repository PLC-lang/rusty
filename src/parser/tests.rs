use crate::ast::PrimitiveType;
use crate::ast::Type;
use crate::ast::PouType;
use crate::ast::VariableBlockType;
use crate::lexer;
use pretty_assertions::*;

#[test]
fn empty_returns_empty_compilation_unit() {
    let result = super::parse(lexer::lex("")).unwrap();
    assert_eq!(result.units.len(), 0);
}

#[test]
fn empty_global_vars_can_be_parsed() {
    let lexer = lexer::lex("VAR_GLOBAL END_VAR");
    let result = super::parse(lexer).unwrap();

    let vars = &result.global_vars[0]; //globar_vars
    let ast_string = format!("{:#?}", vars);
    let expected_ast = 
r#"VariableBlock {
    variables: [],
    variable_block_type: Global,
}"#;
    assert_eq!(ast_string,expected_ast)

}

#[test]
fn global_vars_can_be_parsed() {
    let lexer = lexer::lex("VAR_GLOBAL x : INT; y : BOOL; END_VAR");
    let result = super::parse(lexer).unwrap();

    let vars = &result.global_vars[0]; //globar_vars
    let ast_string = format!("{:#?}", vars);
    let expected_ast = 
r#"VariableBlock {
    variables: [
        Variable {
            name: "x",
            data_type: Primitive(
                Int,
            ),
        },
        Variable {
            name: "y",
            data_type: Primitive(
                Bool,
            ),
        },
    ],
    variable_block_type: Global,
}"#;
    assert_eq!(ast_string,expected_ast)

}

#[test]
fn two_global_vars_can_be_parsed() {
    let lexer = lexer::lex("VAR_GLOBAL a: INT; END_VAR VAR_GLOBAL x : INT; y : BOOL; END_VAR");
    let result = super::parse(lexer).unwrap();

    let vars = &result.global_vars; //globar_vars
    let ast_string = format!("{:#?}", vars);
    let expected_ast = 
r#"[
    VariableBlock {
        variables: [
            Variable {
                name: "a",
                data_type: Primitive(
                    Int,
                ),
            },
        ],
        variable_block_type: Global,
    },
    VariableBlock {
        variables: [
            Variable {
                name: "x",
                data_type: Primitive(
                    Int,
                ),
            },
            Variable {
                name: "y",
                data_type: Primitive(
                    Bool,
                ),
            },
        ],
        variable_block_type: Global,
    },
]"#;
    assert_eq!(ast_string,expected_ast)

}

#[test]
fn simple_foo_program_can_be_parsed() {
    let lexer = lexer::lex("PROGRAM foo END_PROGRAM");
    let result = super::parse(lexer).unwrap();

    let prg = &result.units[0];
    assert_eq!(prg.pou_type, PouType::Program);
    assert_eq!(prg.name, "foo");
    assert!(prg.return_type.is_none());
}

#[test]
fn simple_foo_function_can_be_parsed() {
    let lexer = lexer::lex("FUNCTION foo : INT END_FUNCTION");
    let result = super::parse(lexer).unwrap();

    let prg = &result.units[0];
    assert_eq!(prg.pou_type, PouType::Function);
    assert_eq!(prg.name, "foo");
    assert_eq!(prg.return_type.as_ref().unwrap(), &Type::Primitive(PrimitiveType::Int));
}

#[test]
fn simple_foo_function_block_can_be_parsed() {
    let lexer = lexer::lex("FUNCTION_BLOCK foo END_FUNCTION_BLOCK");
    let result = super::parse(lexer).unwrap();

    let prg = &result.units[0];
    assert_eq!(prg.pou_type, PouType::FunctionBlock);
    assert_eq!(prg.name, "foo");
    assert!(prg.return_type.is_none());
}

#[test]
fn two_programs_can_be_parsed() {
    let lexer = lexer::lex("PROGRAM foo END_PROGRAM  PROGRAM bar END_PROGRAM");
    let result = super::parse(lexer).unwrap();

    let prg = &result.units[0];
    assert_eq!(prg.name, "foo");
    let prg2 = &result.units[1];
    assert_eq!(prg2.name, "bar");
}

#[test]
fn simple_program_with_varblock_can_be_parsed() {
    let lexer = lexer::lex("PROGRAM buz VAR END_VAR END_PROGRAM");
    let result = super::parse(lexer).unwrap();

    let prg = &result.units[0];

    assert_eq!(prg.variable_blocks.len(), 1);
}

#[test]
fn simple_program_with_two_varblocks_can_be_parsed() {
    let lexer = lexer::lex("PROGRAM buz VAR END_VAR VAR END_VAR END_PROGRAM");
    let result = super::parse(lexer).unwrap();

    let prg = &result.units[0];

    assert_eq!(prg.variable_blocks.len(), 2);
}

#[test]
fn a_program_needs_to_end_with_end_program() {
    let lexer = lexer::lex("PROGRAM buz ");
    let result = super::parse(lexer);
    assert_eq!(result, Err("unexpected end of body End, statements : []".to_string()));
}

#[test]
fn a_variable_declaration_block_needs_to_end_with_endvar() {
    let lexer = lexer::lex("PROGRAM buz VAR END_PROGRAM ");
    let result = super::parse(lexer);
    assert_eq!(
        result,
        Err("expected KeywordEndVar, but found KeywordEndProgram".to_string())
    );
}


#[test]
fn a_statement_without_a_semicolon_fails() {
    let lexer = lexer::lex("PROGRAM buz x END_PROGRAM ");
    let result = super::parse(lexer);
    assert_eq!(
        result,
        Err("expected End Statement, but found KeywordEndProgram".to_string())
    );
}

#[test]
fn empty_statements_are_ignored() {
    let lexer = lexer::lex("PROGRAM buz ;;;; END_PROGRAM ");
    let result = super::parse(lexer).unwrap();
    
    let prg = &result.units[0];
    assert_eq!(0, prg.statements.len());
}

#[test]
fn empty_statements_are_ignored_before_a_statement() {
    let lexer = lexer::lex("PROGRAM buz ;;;;x; END_PROGRAM ");
    let result = super::parse(lexer).unwrap();
    
    let prg = &result.units[0];
    let statement = &prg.statements[0];

    let ast_string = format!("{:#?}", statement);
    let expected_ast = r#"Reference {
    name: "x",
}"#;
    assert_eq!(ast_string, expected_ast);
}

#[test]
fn empty_statements_are_ignored_after_a_statement() {
    let lexer = lexer::lex("PROGRAM buz x;;;; END_PROGRAM ");
    let result = super::parse(lexer).unwrap();
    
    let prg = &result.units[0];
    let statement = &prg.statements[0];

    let ast_string = format!("{:#?}", statement);
    let expected_ast = r#"Reference {
    name: "x",
}"#;
    assert_eq!(ast_string, expected_ast);
}

#[test]
fn simple_program_with_variable_can_be_parsed() {
    let lexer = lexer::lex("PROGRAM buz VAR x : INT; END_VAR END_PROGRAM");
    let result = super::parse(lexer).unwrap();

    let prg = &result.units[0];
    let variable_block = &prg.variable_blocks[0];
    let ast_string = format!("{:#?}", variable_block);
    let expected_ast = 
r#"VariableBlock {
    variables: [
        Variable {
            name: "x",
            data_type: Primitive(
                Int,
            ),
        },
    ],
    variable_block_type: Local,
}"#;
    assert_eq!(ast_string,expected_ast);

}

#[test]
fn simple_program_with_var_input_can_be_parsed() {
    
    let lexer = lexer::lex("PROGRAM buz VAR_INPUT x : INT; END_VAR END_PROGRAM");
    let result = super::parse(lexer).unwrap();

    let prg = &result.units[0];
    let variable_block = &prg.variable_blocks[0];
    let ast_string = format!("{:#?}", variable_block);
    let expected_ast = 
r#"VariableBlock {
    variables: [
        Variable {
            name: "x",
            data_type: Primitive(
                Int,
            ),
        },
    ],
    variable_block_type: Input,
}"#;
    assert_eq!(ast_string,expected_ast);
}
