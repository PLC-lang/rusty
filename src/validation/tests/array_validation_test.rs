use crate::assert_validation_snapshot;
use crate::test_utils::tests::parse_and_validate;

#[test]
fn array_access_validation() {
    let diagnostics = parse_and_validate(
        "
			VAR_GLOBAL CONSTANT
				start : INT := 1;
				end : INT := 2;
			END_VAR

        	PROGRAM prg
        	VAR
				multi : ARRAY[0..1,2..3] OF INT;
				nested : ARRAY[0..1] OF ARRAY[2..3] OF INT;
				arr : ARRAY[0..1] OF INT;
				negative_start : ARRAY[-2..2] OF INT;
				negative : ARRAY[-3..-1] OF INT;
				const : ARRAY[start..end] OF INT;
				int_ref : INT;
				string_ref : STRING;
        	END_VAR

			// valid
			multi[0,3];
			nested[1][3];
			arr[1];
			negative_start[-1];
			negative[-2];
			const[1];
			arr[int_ref];

			// invalid
			multi[1,4]; // out of range
			nested[1][4]; // out of range
			arr[3]; // out of range
			negative_start[-4]; // out of range
			negative[-4]; // out of range
			const[3]; // out of range
			arr[string_ref]; // invalid type for array access
			int_ref[1]; // not an array
        	END_PROGRAM
       ",
    );

    assert_validation_snapshot!(&diagnostics);
}

#[test]
fn array_initialization_validation() {
    let diagnostics = parse_and_validate(
        "
		FUNCTION main : DINT
		VAR
			arr			: ARRAY[1..2] OF DINT;
			arr2		: ARRAY[1..2] OF DINT := 1, 2; 												// Missing `[`
			arr3		: ARRAY[1..2] OF myStruct := ((var1 := 1), (var1 := 2, var2 := (1, 2))); 	// Missing `[`
			arr4		: ARRAY[1..2] OF myStruct := ((var1 := 1), (var1 := 2, var2 := 1, 2)); 		// Missing `[`
			arr_init	: ARRAY[1..2] OF DINT := (1, 2);											// Missing `[`
			x	 		: myStruct;
			y	 		: myStruct := (var1 := 1, var2 := 3, 4);									// Missing `[`
		END_VAR
			arr	:= 1, 2; 																			// Missing `[`
			arr	:= (1, 2);																			// Missing `[`
			arr	:= (arr_init); // valid
			x	:= (var1 := 1, var2 := 3, 4); 														// Missing `[`
			x	:= (var1 := 1, var2 := arr_init); // valid
		END_FUNCTION
		
		TYPE myStruct : STRUCT
				var1 : DINT;
				var2 : ARRAY[1..2] OF DINT;
			END_STRUCT
		END_TYPE
       ",
    );

    assert_validation_snapshot!(&diagnostics);
}

#[test]
fn array_access_dimension_mismatch() {
    let diagnostics = parse_and_validate(
        "
		FUNCTION fn : DINT
			VAR_INPUT {ref}
				arr : ARRAY[0..5] OF DINT;
				vla : ARRAY[*] OF DINT;
			END_VAR

			// Valid
			arr[0] := 1;
			vla[0] := 1;

			// Invalid
			arr[0, 1] := 1;
			vla[0, 1] := 1;
			arr[0, 1, 2] := 1;
			vla[0, 1, 2] := 1;
		END_FUNCTION
		",
    );

    assert_eq!(diagnostics.len(), 4);
    assert_validation_snapshot!(diagnostics);
}

#[test]
fn assignment_1d() {
    let diagnostics = parse_and_validate(
        "
		FUNCTION main : DINT
			VAR
				arr 		: ARRAY[1..5] OF DINT := [1, 2, 3, 4, 5, 6];
				arr_alt		: ARRAY[1..5] OF DINT := (1, 2, 3, 4, 5, 6);
			END_VAR

			arr := [1, 2, 3, 4];	// Valid
			arr := [1, 2, 3, 4, 5]; // Valid

			arr := (1, 2, 3, 4);
			arr := (1, 2, 3, 4, 5);
			arr := (1, 2, 3, 4, 5, 6);
			arr := [1, 2, 3, 4, 5, 6];
			arr := [1, 2, 3, 4, 5);
			arr := (1, 2, 3, 4, 5];
		END_FUNCTION
		",
    );

    assert_validation_snapshot!(diagnostics);
}

#[test]
fn assignment_2d() {
    let diagnostics = parse_and_validate(
        "
		FUNCTION main : DINT
			VAR
				arr 			: ARRAY[1..2, 1..5] OF DINT := [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11];
				arr_alt			: ARRAY[1..2, 1..5] OF DINT := (1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11);

				arr_nested		: ARRAY[1..2] OF ARRAY[1..5] OF DINT := [ [1, 2, 3, 4, 5], [6, 7, 8, 9, 10], [11, 12, 13, 14, 15] ];
				arr_nested_alt	: ARRAY[1..2] OF ARRAY[1..5] OF DINT := ( [1, 2, 3, 4, 5], [6, 7, 8, 9, 10], [11, 12, 13, 14, 15] );
			END_VAR

			arr := [1, 2, 3, 4, 5, 6, 7, 8, 9]; 	// Valid
			arr := [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]; // Valid

			arr := (1, 2, 3, 4, 5, 6, 7, 8, 9);
			arr := (1, 2, 3, 4, 5, 6, 7, 8, 9, 10);
			arr := (1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11);
			arr := [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11];
			
			arr_nested		:= [ [1, 2, 3, 4, 5], [6, 7, 8, 9, 10] ]; // Valid
			arr_nested		:= [ [1, 2, 3, 4, 5], [6, 7, 8, 9, 10], [11, 12, 13, 14, 15] ];
			arr_nested		:= ( [1, 2, 3, 4, 5], [6, 7, 8, 9, 10], [11, 12, 13, 14, 15] );
		END_FUNCTION
		",
    );

    assert_validation_snapshot!(diagnostics);
}

#[test]
fn assignment_3d() {
    let diagnostics = parse_and_validate(
        "
		FUNCTION main : DINT
			VAR
				arr 		: ARRAY[1..2, 1..2, 1..2] OF DINT := [1, 2, 3, 4, 5, 6, 7, 8, 9];
				arr_alt 	: ARRAY[1..2, 1..2, 1..2] OF DINT := (1, 2, 3, 4, 5, 6, 7, 8, 9);

				arr_nested		: ARRAY[1..2] OF ARRAY[1..2] OF ARRAY[1..2] OF DINT := [ [[1, 2], [3, 4]], [[5, 6], [7, 8]], [[9, 10], [11, 12]] ];
				arr_nested_alt	: ARRAY[1..2] OF ARRAY[1..2] OF ARRAY[1..2] OF DINT := ( [[1, 2], [3, 4]], [[5, 6], [7, 8]], [[9, 10], [11, 12]] );
			END_VAR

			arr := [1, 2, 3, 4, 5, 6, 7]; 		// Valid
			arr := [1, 2, 3, 4, 5, 6, 7, 8]; 	// Valid

			arr := (1, 2, 3, 4, 5, 6, 7, 8);
			arr := (1, 2, 3, 4, 5, 6, 7, 8, 9);
			arr := [1, 2, 3, 4, 5, 6, 7, 8, 9];
			
			arr_nested := [ [[1, 2], [3, 4]], [[5, 6], [7, 8]] ]; // Valid
			arr_nested := [ [[1, 2], [3, 4]], [[5, 6], [7, 8]], [[9, 10], [11, 12]] ];
			arr_nested := ( [[1, 2], [3, 4]], [[5, 6], [7, 8]], [[9, 10], [11, 12]] );
		END_FUNCTION
		",
    );

    assert_validation_snapshot!(diagnostics);
}

#[test]
fn assignment_structs() {
    let diagnostics = parse_and_validate(
        "
		TYPE FOO : STRUCT
			idx : DINT;
			arr : ARRAY[1..2] OF BAR;
		END_STRUCT END_TYPE

		TYPE BAR : STRUCT
			arr : ARRAY[1..5] OF DINT;
		END_STRUCT END_TYPE

		FUNCTION main : DINT
			VAR
				foo_valid_0 	: FOO := (idx := 0, arr := [(arr := [1, 2, 3, 4, 5])]);
				foo_valid_1 	: FOO := (idx := 0, arr := [(arr := [1, 2, 3, 4, 5]), (arr := [1, 2, 3])]);
				foo_valid_2 	: FOO := (idx := 0, arr := [(arr := [1, 2, 3, 4, 5]), (arr := [1, 2, 3, 4, 5])]);
				foo_valid_3 	: FOO := (arr := [(arr := [1, 2, 3, 4, 5]), (arr := [1, 2, 3, 4, 5])], idx := 0);

				foo_invalid_0 	: FOO := (idx := 0, arr := ((arr := (1, 2, 3, 4, 5)), (arr := (1, 2, 3, 4, 5))));
				foo_invalid_1 	: FOO := (idx := 0, arr := ((arr := (1, 2, 3, 4, 5)), (arr := (1, 2, 3, 4, 5))));
			END_VAR
		END_FUNCTION
		",
    );

    assert_validation_snapshot!(diagnostics);
}

#[test]
#[ignore = "Needs to be re-checked"]
fn exceeding_size_structs() {
    let diagnostics = parse_and_validate(
        "
		TYPE Foo : STRUCT
			idx: DINT;
			arr : ARRAY[1..2] OF Bar;
		END_STRUCT END_TYPE

		TYPE Bar : STRUCT
			arr : ARRAY[1..2] OF DINT;
		END_STRUCT END_TYPE

		FUNCTION main : DINT
			VAR
				arr_a : Foo := (
					idx := 1, 
					arr := [(arr := [1, 2]), (arr := [3, 4]), (arr := [5, 6])] 			// Invalid, because the outter `arr` can only store 2 elements
				);

				arr_b : ARRAY[1..2] OF Foo := (
					[idx := 2, arr := [(arr := [1, 2, 3]),  (arr := [4, 5])]],			// Invalid because of the first inner `arr`
					[idx := 3, arr := [(arr := [1, 2]),     (arr := [3, 4, 5])]],		// Invalid because of the second inner `arr`
				);

				arr_c : ARRAY[1..2] OF Foo := (											// Invalid because only 2 elements can be stored, but 3 are provided
					(idx := 4, arr := [(arr := [1, 2, 3]),  (arr := [4, 5])]),			// Invalid because of the first inner `arr`
					(idx := 5, arr := [(arr := [1, 2]),     (arr := [3, 4, 5])]),		// Invalid because of the second inner `arr`
					(idx := 6, arr := [(arr := 2(0)),  		(arr := 4(0))]),			// Invalid ebcause of the second inner `arr`
				);
			END_VAR
		END_FUNCTION
		",
    );

    assert_validation_snapshot!(diagnostics)
}

#[test]
fn assignment_multiplied_statement() {
    let diagnostics = parse_and_validate(
        "
		FUNCTION main : DINT
			VAR
				arr_1d : ARRAY[1..5] OF DINT := [6(0)];

				arr_2d : ARRAY[1..2, 1..5] OF DINT := [11(0)];
				arr_2d_nested : ARRAY[1..2] OF ARRAY[1..5] OF DINT := [11(0)];

				arr_3d : ARRAY[1..2, 1..2, 1..2] OF DINT := [9(0)];
				arr_3d_nested : ARRAY[1..2] OF ARRAY[1..2] OF ARRAY[1..2] OF DINT := [9(0)];
			END_VAR

			// Valid
			arr_1d := [5(0)];
			arr_2d := [10(0)];
			arr_2d_nested := [10(0)];
			arr_3d = [8(0)];
			arr_3d_nested := [8(0)];

			// Invalid
			arr_1d := [6(0)];
			arr_2d := [11(0)];
			arr_2d_nested := [11(0)];
			arr_3d := [9(0)];
			arr_3d_nested := [9(0)];
		END_FUNCTION
		",
    );

    assert_validation_snapshot!(diagnostics);
}
