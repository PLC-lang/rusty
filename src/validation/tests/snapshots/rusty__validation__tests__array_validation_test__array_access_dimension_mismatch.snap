---
source: src/validation/tests/array_validation_test.rs
expression: diagnostics
---
error: Expected array access with 1 dimensions, found 2
   ┌─ <internal>:13:17
   │
13 │             arr[0, 1] := 1;
   │                 ^^^^ Expected array access with 1 dimensions, found 2

error: Expected array access with 1 dimensions, found 2
   ┌─ <internal>:14:17
   │
14 │             vla[0, 1] := 1;
   │                 ^^^^ Expected array access with 1 dimensions, found 2

error: Expected array access with 1 dimensions, found 3
   ┌─ <internal>:15:17
   │
15 │             arr[0, 1, 2] := 1;
   │                 ^^^^^^^ Expected array access with 1 dimensions, found 3

error: Expected array access with 1 dimensions, found 3
   ┌─ <internal>:16:17
   │
16 │             vla[0, 1, 2] := 1;
   │                 ^^^^^^^ Expected array access with 1 dimensions, found 3


